src/devicecontroller/DefaultDeviceController.ts

```diff
@@ -496,30 +503,36 @@ export default class DefaultDeviceController implements DeviceControllerBasedMed
       this.logger.info(`reusing existing ${kind} device`);
       return DevicePermission.PermissionGrantedPreviously;
     }
-    const oldStream: MediaStream | null = this.activeDevices[kind]
-      ? this.activeDevices[kind].stream
-      : null;
-    if (kind === 'audio') {
-      this.releaseMediaStream(oldStream);
-    }
-    let startTimeMs = Date.now();
-    let newDevice: DeviceSelection;
+
+    const startTimeMs = Date.now();
+    const newDevice: DeviceSelection = new DeviceSelection();
     try {
       this.logger.info(
         `requesting new ${kind} device with constraint ${JSON.stringify(proposedConstraints)}`
       );
       const stream = this.deviceAsMediaStream(device);
       if (kind === 'audio' && device === null) {
-        newDevice = new DeviceSelection();
         newDevice.stream = DefaultDeviceController.createEmptyAudioDevice() as MediaStream;
         newDevice.constraints = null;
       } else if (stream) {
         this.logger.info(`using media stream ${stream.id} for ${kind} device`);
-        newDevice = new DeviceSelection();
         newDevice.stream = stream;
         newDevice.constraints = proposedConstraints;
       } else {
-        newDevice = await this.getDeviceFromBrowser(proposedConstraints);
+        newDevice.stream = await navigator.mediaDevices.getUserMedia(proposedConstraints);
+        newDevice.constraints = proposedConstraints;
+
+        if (kind === 'video' && this.lastNoVideoInputDeviceCount > callCount) {
```

Comment:
When can this.lastNoVideoInputDeviceCount > callCount? I saw they all is set to this.videoInputDeviceCount.

Reply:
1. Call `chooseVideoInputDevice()` and wait for `getUserMedia()` to resolve. A user needs to click Allow or Deny, or it will take a few milliseconds if previously allowed.
2. While waiting for `getUserMedia()`, call `chooseVideoInputDevice(null)` to unselect all devices.
3. Now, the previous call at (1) is resolved.
   At this point, `this.lastNoVideoInputDeviceCount > callCount` will be `true`.